src/seshat.erl

```diff
@@ -2,175 +2,380 @@
+%% @doc Return the metadata for the fields
+%% When creating a set of metrics with seshat:new/3 or seshat:new/4,
+%% metadata about the metrics/counters has to be provided either
+%% as a list or as a reference to a persistent term with the list.
+%% The latter is recommended when a lot of similar metrics need to be stored
+%% (eg. many instances of the same component emit the same set of metrics)
+-spec resolve_fields_spec(fields_spec()) -> [field_spec()].
+resolve_fields_spec(Fields = FieldsSpec) when is_list(FieldsSpec) ->
     Fields;
-resolve_fields({persistent_term, PTerm}) ->
+resolve_fields_spec({persistent_term, PTerm}) ->
     %% TODO error handling
     persistent_term:get(PTerm).
 
-register_counter(Group, Name, Ref, Fields) ->
+-spec register_counter(group(), id(), counters:counters_ref(), fields_spec(), labels()) ->
+    ok.
+register_counter(Group, Id, CRef, FieldsSpec, Labels) when is_map(Labels) ->
     TRef = seshat_counters_server:get_table(Group),
-    true = ets:insert(TRef, {Name, Ref, Fields}),
+    Entry = #entry{
+        id = Id,
+        cref = CRef,
+        field_spec = FieldsSpec,
+        labels = Labels,
+        rendered_labels = labels_to_binary(Labels)
+    },
+    true = ets:insert(TRef, Entry),
     ok.
 
-new_counter(Group, Name, Fields, FieldsSpec) ->
+-spec new_counter(group(), id(), [field_spec()], fields_spec(), labels()) ->
+    counters:counters_ref().
+new_counter(Group, Id, Fields, FieldsSpec, Labels) ->
     Size = length(Fields),
-    %% TODO: validate that positions are correct, i.e. not out of range
-    %% or duplicated
-    ExpectedPositions = lists:seq(1, Size),
-    Positions = lists:sort([P || {_, P, _, _} <- Fields]),
-    case ExpectedPositions == Positions of
+    ExpectedIndexes = lists:seq(1, Size),
+    Indexes = lists:sort([P || {_, P, _, _} <- Fields]),
+    case ExpectedIndexes == Indexes of
         true ->
             CRef = counters:new(Size, [write_concurrency]),
-            ok = register_counter(Group, Name, CRef, FieldsSpec),
+            ok = register_counter(Group, Id, CRef, FieldsSpec, Labels),
             CRef;
         false ->
             error(invalid_field_specification)
     end.
 
+%% @doc Return a Prometheus-formated text (as a binary),
+%% which can be directly returned by a Prometheus endpoint.
+%% The returned binary has the following structure:
+%% prefix_name{label1="value1",...} Value
+%%
+%% @param Group the name of an existing group
+%% @param Names the list of metrics to return
+%%
+-spec prom_format(group(), string()) -> binary().
+prom_format(Group, Prefix) ->
+    do_prom_format(format(Group, #{metrics => all, labels => as_binary}), Prefix).
+
+-spec prom_format(group(), string(), [atom()]) -> binary().
+prom_format(Group, Prefix, Names) when is_list(Names) ->
+    do_prom_format(format(Group, #{metrics => Names, labels => as_binary}), Prefix).
+
+-spec do_prom_format(format_result(), string()) -> binary().
+do_prom_format(Data, Prefix) ->
```

Comment:
I'm thinking the api should return iodata() isntead of and you can aggregate the lines using lists which should be more efficient.